### PR TITLE
Hoist exits out of loops when unrolling to make code structured.

### PR DESCRIPTION
**Pull request by adam-yang**

### CODE REVIEW

File: lib/Transforms/Scalar/LoopUnrollPass.cpp [hunk 1]
```diff
@@ -933,6 +935,8 @@ bool LoopUnroll::runOnLoop(Loop *L, LPPassManager &LPM) {
     return false;
   }
 
+  hlsl::RemoveUnstructuredLoopExits(L, LI, &getAnalysis<DominatorTreeWrapperPass>().getDomTree()); // HLSL Change
```
[the remark this answers is not in the document]
Comment (adam-yang) [reply lib/Transforms/Scalar/LoopUnrollPass.cpp:938]: It'll only make change if it detects multiple exit blocks.

[the remark this answers is not in the document]
Comment (python3kgae) [reply lib/Transforms/Scalar/LoopUnrollPass.cpp:938]: Could we only do it when unroll? If not unroll, will it cause unstructured flow?

File: lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+//        if (!broke) {
+//          code_1;
+//        }
+//
+//        if (exit || broke)
+//          break;
+//      }
+//
+//      if (broke) {
+//        exit_code;
```
[the remark this answers is not in the document]
Comment (adam-yang) [reply lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp:51]: Each break gets a flag.

File: lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+static BasicBlock *GetExitBockForExitingBlock(Loop *L, BasicBlock *exiting_block) {
+  BranchInst *br = dyn_cast<BranchInst>(exiting_block->getTerminator());
+  assert(L->contains(exiting_block));
+  assert(br->isConditional());
+  BasicBlock *result = L->contains(br->getSuccessor(0)) ? br->getSuccessor(1) : br->getSuccessor(0);
+  assert(!L->contains(result));
+  return result;
+}
+
+static BasicBlock::iterator FindFirstNonPhi(BasicBlock *bb) {
```
[the remark this answers is not in the document]
Comment (adam-yang) [reply lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp:116]: Will do..

File: lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (adam-yang) [reply lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp:51]: The break flags for different exits are mutually exclusive.

File: lib/Transforms/Scalar/LoopUnrollPass.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (adam-yang) [reply lib/Transforms/Scalar/LoopUnrollPass.cpp:938]: Yeah this basically only runs if it unrolls. At this point we expect the loop to be unrolled. The UnrollLoop function returns false if there's malformed loop (multiple latch, etc.) or something.

File: lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (python3kgae) [reply lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp:51]: Could we update the example to have more breaks which need flag?

[the remark this answers is not in the document]
Comment (adam-yang) [reply lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp]: Did it.

File: lib/Transforms/Utils/SimplifyCFG.cpp [hunk 4]
```diff
@@ -4573,12 +4573,14 @@ bool SimplifyCFGOpt::SimplifyCondBranch(BranchInst *BI, IRBuilder<> &Builder) {
         return SimplifyCFG(BB, TTI, BonusInstThreshold, AC) | true;
   }
 
+#if 0 // HLSL Change - This often creates unstructured flow
```
[the remark this answers is not in the document]
Comment (adam-yang) [reply lib/Transforms/Utils/SimplifyCFG.cpp:4578]: It's hard to detect, I'll measure perf to see the impact.

**Comment by pow2clk** [comment]
Do you think we might want an option to enable or disable this? If so, you could use the -opt-enable or -opt-disable flags already present.

File: lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+
+        for (unsigned i = 0; i < phi->getNumIncomingValues(); i++) {
+          if (phi->getIncomingBlock(i) == bb) {
+            phi->setIncomingValue(i, data.value);
+            break;
+          }
+        }
+
+        if (!seen.count(succ)) {
+          work_list.push_back({ succ, phi });
```
Comment (pow2clk) [inline lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp:264]: I've confirmed that changing this to `{ succ, phi, false }` is all that remains to get the Linux build working.

Comment (adam-yang) [reply lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp:264]: thank you.

File: lib/Transforms/Utils/SimplifyCFG.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (adam-yang) [reply lib/Transforms/Utils/SimplifyCFG.cpp]: No longer doing this. It actually caused issues.

File: lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+//      for(;;) {
+//        if (a) {
+//          if (b) {
+//            broke_0 = true;
+//          }
+//
+//          if (!broke_0) {
+//            if (c) {
+//              if (d) {
+//                exit_code_1;
```
Comment (adam-yang) [inline lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp:56]: delete this line.

File: lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+//              code_2;
+//            }
+//          }
+//
+//          ...
+//        }
+//
+//        if (!broke_0) {
+//          if (!broke_1) {
+//            code_1;
```
Comment (adam-yang) [inline lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp:74]: should be code 3

File: lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+      // Do not include the exiting block itself in this calculation
+      if (!data.is_exiting_block) {
+        // If this block is part of an inner loop... Give up for now.
+        if (LI->getLoopFor(data.bb) != L) {
+          give_up = true;
+        }
+        // Otherwise just remember the blocks with side effects (including the latch)
+        else {
+          for (Instruction &I : *bb) {
+            if (I.mayWriteToMemory() && !IsNoop(&I)) {
```
Comment (adam-yang) [inline lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp:229]: Don't allow read memory either.

File: lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+
+  for (;;) {
+    // Recompute exiting block every time, since they could change between
+    // iterations
+    llvm::SmallVector<BasicBlock *, 4> exiting_blocks;
+    L->getExitingBlocks(exiting_blocks);
+
+    bool local_changed = false;
+
+    // Don't bother if this exit branch already dominates latch (or is the latch)
```
Comment (adam-yang) [inline lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp:403]: Get rid of this comment.

File: lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp [hunk 10]
```diff
@@ -0,0 +1,427 @@
...
+  bool give_up = false;
+  std::unordered_map<BasicBlock *, PHINode *> cached_phis;
+
+  // Use a worklist to propagate the exit condition from within the block
+  {
+    Value *false_value = ConstantInt::getFalse(i1Ty);
+
+    struct Propagate_Data {
+      BasicBlock *bb;
+      Value *value;
```
Comment (adam-yang) [inline lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp:203]: `exit_condition` maybe

File: lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+
+    BasicBlock *latch = L->getLoopLatch();
+
+    for (unsigned i = 0; i < work_list.size(); i++) {
+      Propagate_Data data = work_list[i];
+
+      BasicBlock *bb = data.bb;
+
+      // Do not include the exiting block itself in this calculation
+      if (!data.is_exiting_block) {
```
Comment (adam-yang) [inline lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp:221]: is_first, maybe. Or could use `i != 0`

File: tools/clang/test/HLSLFileCheck/hlsl/control_flow/loops/struct_exit_outer_loop.hlsl [hunk 12]
```diff
@@ -0,0 +1,99 @@
+// RUN: %dxc -E main -Zi -O3 -T ps_6_0 %s | FileCheck %s
+// RUN: %dxc -E main -Zi -T ps_6_0 %s | FileCheck %s
```
Comment (adam-yang) [inline tools/clang/test/HLSLFileCheck/hlsl/control_flow/loops/struct_exit_outer_loop.hlsl:2]: Zi.

File: tools/clang/test/HLSLFileCheck/hlsl/control_flow/loops/struct_exit_exit_value.hlsl [hunk 13]
```diff
@@ -0,0 +1,62 @@
+// RUN: %dxc -E main -Od -T ps_6_0 %s -DFORCE_UNROLL | FileCheck %s
```
Comment (adam-yang) [inline tools/clang/test/HLSLFileCheck/hlsl/control_flow/loops/struct_exit_exit_value.hlsl:1]: Zi

**Review by pow2clk — APPROVED** [review]
Latest changes look good!

**Comment by pow2clk** [comment]
There's something hinky about that ubuntu failure. It seems to have timed out waiting to clone after 2 hours. 

The travis build succeeded, https://travis-ci.org/github/microsoft/DirectXShaderCompiler/builds/723938384

**Comment by pow2clk** [comment]
I think you should go ahead and merge it.

**Pull request merged**